File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/5. 15647.py
import sys
from array import array
from collections import deque

input = sys.stdin.readline
output = sys.stdout.write

# 입력 저장
N = int(input())
connInfo = [[] for _ in range(N + 1)]

for _ in range(N - 1):
  u, v, w = map(int, input().split())
  connInfo[u].append((v, w))
  connInfo[v].append((u, w))

# array 모듈로 메모리 절약 (C 스타일 정수)
nowSum = array('q', [0] * (N + 1))
numNode = array('q', [0] * (N + 1))
parentArr = array('q', [-1] * (N + 1))
resultArr = array('q', [0] * (N + 1))

rootNode = 1
parentArr[rootNode] = 0

# BFS로 부모·자식 관계 구축
children = [[] for _ in range(N + 1)]
visitQueue = deque([rootNode])
while visitQueue:
  now = visitQueue.popleft()
  for nextNode, dist in connInfo[now]:
    if parentArr[nextNode] == -1:
      parentArr[nextNode] = now
      children[now].append((nextNode, dist))
      visitQueue.append(nextNode)

# 역순 레벨 순서 (리프 → 루트)
order = []
visitQueue.append(rootNode)
while visitQueue:
  now = visitQueue.popleft()
  order.append(now)
  for nextNode, _ in children[now]:
    visitQueue.append(nextNode)

# dfs_1: 리프부터 처리
for now in reversed(order):
  nowSumVal = 0
  nowNodes = 1
  for nextNode, dist in children[now]:
    nowSumVal += nowSum[nextNode] + dist * numNode[nextNode]
    nowNodes += numNode[nextNode]
  nowSum[now] = nowSumVal
  numNode[now] = nowNodes

# dfs_2: 루트부터 처리
visitQueue.append((rootNode, 0, 0, 0, 0))
while visitQueue:
  now, argSumNow, argSumAcc, argSumLen, distParent = visitQueue.popleft()
  resultArr[now] = nowSum[now] + argSumNow + argSumAcc + argSumLen * distParent

  sumNow = argSumNow
  sumAcc = argSumAcc
  sumLen = argSumLen + 1
  for nextNode, dist in children[now]:
    sumNow += nowSum[nextNode]
    sumAcc += numNode[nextNode] * dist
    sumLen += numNode[nextNode]
  sumAcc += argSumLen * distParent

  for nextNode, dist in children[now]:
    nextSumNow = sumNow - nowSum[nextNode]
    nextSumAcc = sumAcc - numNode[nextNode] * dist
    nextSumLen = sumLen - numNode[nextNode]
    visitQueue.append((nextNode, nextSumNow, nextSumAcc, nextSumLen, dist))

for i in range(1, N + 1):
  output(str(resultArr[i]) + '\n')

File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/5. 15647.js
// 입력 모듈
const fs = require('fs')

// 입력 저장
let N = 0
const connInfoArr = Array.from({length: 300001}, () => [])

fs.readFileSync(0)
  .toString()
  .trim()
  .split('\n')
  .forEach((inputRow, rowIdx) => {
    if (rowIdx === 0) {
      N = +inputRow
    } // ::
    else {
      const [U, V, W] = inputRow.trim().split(' ').map(Number)
      connInfoArr[U].push([V, W])
      connInfoArr[V].push([U, W])
    }
  })

/**
 * nowSumArr[i]: i 가 루트일때, 자손들과의 거리합
 * numNodeArr[i]: i 가 루트인 서브트리의 노드 개수(본인 포함)
 * parentArr[i]: i 의 부모노드 번호. 루트의 부모는 0, 설정 안됬으면 -1
 * resultArr[i]: 노드 i 에서 나머지 노드들까지의 거리합
 */
const nowSumArr = Array.from({length: N + 1}).fill(0)
const numNodeArr = Array.from({length: N + 1}).fill(0)
const parentArr = Array.from({length: N + 1}).fill(-1)
const resultArr = Array.from({length: N + 1}).fill(0)

const ROOT_NODE = 1
parentArr[ROOT_NODE] = 0

DFS_1(ROOT_NODE)
DFS_2(ROOT_NODE, 0, 0, 0, 0)

let resultStr = ''
resultArr.slice(1).forEach(res => (resultStr += res + '\n'))
process.stdout.write(resultStr)

function DFS_1(now) {
  let nowSum = 0
  let nowNodes = 1

  connInfoArr[now].forEach(info => {
    const [next, dist] = info

    if (parentArr[next] === -1) {
      parentArr[next] = now
      const [sum, numNode] = DFS_1(next)

      nowSum += sum + dist * numNode
      nowNodes += numNode
    }
  })

  nowSumArr[now] = nowSum
  numNodeArr[now] = nowNodes

  return [nowSum, nowNodes]
}

function DFS_2(now, argSumNow, argSumAcc, argSumLen, distParent) {
  resultArr[now] = nowSumArr[now] + argSumNow + argSumAcc + argSumLen * distParent

  let sumNow = argSumNow
  let sumAcc = argSumAcc
  let sumLen = argSumLen + 1

  connInfoArr[now].forEach(info => {
    const [next, dist] = info
    if (parentArr[next] === now) {
      sumNow += nowSumArr[next]
      sumAcc += numNodeArr[next] * dist
      sumLen += numNodeArr[next]
    }
  })

  sumAcc += argSumLen * distParent

  connInfoArr[now].forEach(info => {
    const [next, dist] = info
    if (parentArr[next] === now) {
      const nextSumNow = sumNow - nowSumArr[next]
      const nextSumAcc = sumAcc - numNodeArr[next] * dist
      const nextSumLen = sumLen - numNodeArr[next]
      DFS_2(next, nextSumNow, nextSumAcc, nextSumLen, dist)
    }
  })
}
